src/SimplePie.php

```diff
@@ -1901,7 +1901,7 @@ protected function fetch_data(&$cache)
                     $this->data = [];
                 }
                 // Check if the cache has been updated
-                elseif (isset($this->data['cache_expiration_time']) && $this->data['cache_expiration_time'] > time()) {
+                elseif (isset($this->data['cache_expiration_time']) && $this->data['cache_expiration_time'] < time()) {
```

Comment:
```suggestion
                elseif (empty($this->data['cache_expiration_time']) || $this->data['cache_expiration_time'] < time()) {
```

Otherwise you hit the next `else` which believes the cache to be still valid

Reply:
Note that the equality here gives problem when tests are running within the same second, which is the reason for your tests failing.

https://github.com/simplepie/simplepie/blob/9385c54382edec0a768fdbe310c3e580c1b5dc81/tests/Integration/CachingTest.php#L123-L124

You can for instance change to `$defaultMtime = time() - 1`;

See https://github.com/FreshRSS/simplepie/pull/23

Reply:
Replacement PR with additional fixes https://github.com/simplepie/simplepie/pull/883
Feel free to merge in this PR or in master as you prefer

Reply:
> Otherwise you hit the next `else` which believes the cache to be still valid

I would expect that when no expiration time is provided, it cannot really expire so this is correct IMO. Though it [should not actually happen](https://github.com/simplepie/simplepie/pull/883#discussion_r1776094433) if I read the code correctly so the `isset` is there mostly for PHPStan.

Reply:
If `cache_expiration_time` is not set for any reason, I would want my cache to be revalidated. Otherwise it can stay like that for a long time

Reply:
Fair enough. I still wonder why it duplicates the ttl handling, though.